### PR TITLE
fix: handle race condition in terminal compound commands

### LINKED ISSUE #7430: Terminal mishandles compound commands
### App Version

3.25.23

### API Provider

Not Applicable / Other

### Model Used

N/A

### Roo Code Task Links (Optional)

_No response_

### 🔁 Steps to Reproduce

Start a task and make the LLM use a compound terminal command, where the output is generated by the second part. (by compound command I mean a command that uses an additional process spawning operator, such as Linux's `&&`, the most typical example in my cases being `cd dir && run test`)

### 💥 Outcome Summary

The LLM does not see the output of the command. The TerminalRegistry reports an error with no process running for the terminal. Moreover, a new terminal is spawned every time this happens.

### 📄 Relevant Logs or Errors (Optional)

```shell
console.ts:137 [Extension Host] [onDidEndTerminalShellExecution] 
Object
log.ts:460   ERR [Extension Host] [TerminalRegistry] Shell execution end event received, but process is not running for terminal: 
Object
console.ts:137 [Extension Host] [TerminalRegistry] Shell execution end event received, but process is not running for terminal: 
Object
```

### PR DESCRIPTION
**Pull request by daniel-lxs**
## Summary
Fixes #7430 - Terminal mishandles compound commands

## Problem
When executing compound commands (e.g., `cd dir && npm test`) in newly spawned terminals, the LLM would not receive the command output due to a race condition where the shell execution end event arrived before the terminal was marked as running.

## Solution
Modified the `onDidEndTerminalShellExecution` handler in `TerminalRegistry.ts` to properly handle the race condition:
- When `terminal.running = false` but a process exists, we now complete the process anyway instead of returning early
- This ensures the LLM always receives the command output, even when events arrive out of order
- Changed error logging to warning for this specific race condition case

## Changes
- **src/integrations/terminal/TerminalRegistry.ts**: Fixed race condition handling in end event
- **src/integrations/terminal/__tests__/TerminalRegistry.race.spec.ts**: Added comprehensive tests for race condition

## Testing
- ✅ All 45 terminal tests pass (3 new race condition tests + 42 existing)
- ✅ Verified the fix handles the race condition properly
- ✅ No regressions in existing functionality

## How to Test
1. Start a task that uses compound commands like `cd test && echo hello`
2. Verify the LLM receives the output correctly
3. Check logs - should see warning instead of error for race condition
4. No more "process is not running" errors

This fix ensures compound commands work reliably without needing any delays or workarounds.

### CODE REVIEW

File: src/integrations/terminal/TerminalRegistry.ts [hunk 1]
```diff
@@ -95,9 +95,23 @@ export class TerminalRegistry {
...
 					if (!terminal.running) {
+						// Handle race condition: end event arrived before start event could mark terminal as running
+						if (process) {
+							console.warn(
+								"[TerminalRegistry] Shell execution end event received before terminal marked as running (race condition). Completing process anyway:",
+								{ terminalId: terminal?.id, command: process?.command, exitCode: e.exitCode },
+							)
+
+							// Complete the process even though running flag wasn't set
+							terminal.shellExecutionComplete(exitDetails)
```
Comment (ghost) [inline src/integrations/terminal/TerminalRegistry.ts:106]: Is this intentional that we call `terminal.shellExecutionComplete(exitDetails)` here? I'm wondering if there's a risk of duplicate process completion. If the start event eventually arrives and marks the terminal as running, and then another end event arrives, could the process potentially complete twice? Consider adding a flag to track if the process has already been completed to prevent duplicate completions.

File: src/integrations/terminal/TerminalRegistry.ts [hunk 2]
```diff
@@ -95,9 +95,23 @@ export class TerminalRegistry {
 					}
 
 					if (!terminal.running) {
```
Comment (ghost) [inline src/integrations/terminal/TerminalRegistry.ts:97]: Could we consider adding a comment block here explaining why this specific race condition occurs with compound commands and newly spawned terminals? A reference to issue #7430 would help future maintainers understand the context better.

File: src/integrations/terminal/TerminalRegistry.ts [hunk 3]
```diff
@@ -95,9 +95,23 @@ export class TerminalRegistry {
...
+							terminal.shellExecutionComplete(exitDetails)
+							terminal.busy = false
+							return
+						}
+
+						// No process exists - this is an actual error
 						console.error(
-							"[TerminalRegistry] Shell execution end event received, but process is not running for terminal:",
-							{ terminalId: terminal?.id, command: process?.command, exitCode: e.exitCode },
+							"[TerminalRegistry] Shell execution end event received, but process is not running and no process exists:",
```
Comment (ghost) [inline src/integrations/terminal/TerminalRegistry.ts:113]: Minor: The error message format here differs from the warning message format above. Could we make them more consistent? The warning says 'Shell execution end event received before terminal marked as running' while this says 'Shell execution end event received, but process is not running and no process exists'. Consider aligning the phrasing patterns for clarity.

File: src/integrations/terminal/__tests__/TerminalRegistry.race.spec.ts [hunk 4]
```diff
@@ -0,0 +1,205 @@
...
+		mockTerminalInfo = new Terminal(1, mockTerminal, "/test/path")
+		TerminalRegistry["terminals"] = [mockTerminalInfo]
+
+		// Create TerminalProcess
+		terminalProcess = new TerminalProcess(mockTerminalInfo)
+		terminalProcess.command = "cd test && echo hello"
+		mockTerminalInfo.process = terminalProcess
+	})
+
+	it("should handle end event arriving before start event (race condition)", async () => {
```
Comment (ghost) [inline src/integrations/terminal/__tests__/TerminalRegistry.race.spec.ts:92]: Great test coverage for the main race condition scenario! Could we consider adding tests for these edge cases?
- Multiple rapid compound commands executed in sequence
- The start event arriving much later after the race condition has been handled
- The terminal being disposed during the race condition

These would help ensure the fix is robust against various timing scenarios.
